src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -1,42 +1,70 @@
+	vmr "github.com/filecoin-project/specs/systems/filecoin_vm/runtime"
+	exitcode "github.com/filecoin-project/specs/systems/filecoin_vm/runtime/exitcode"
+	gascost "github.com/filecoin-project/specs/systems/filecoin_vm/runtime/gascost"
+	st "github.com/filecoin-project/specs/systems/filecoin_vm/state_tree"
+	util "github.com/filecoin-project/specs/util"
+)
+
+type Bytes = util.Bytes
 
 var TODO = util.TODO
 
-func (vmi *VMInterpreter_I) ApplyMessageBatch(inTree st.StateTree, msgs []MessageRef) (outTree st.StateTree, ret []msg.MessageReceipt) {
-	compTree := inTree
-	for _, m := range msgs {
-		oT, r := vmi.ApplyMessage(compTree, m.Message(), m.Miner())
-		compTree = oT        // assign the current tree. (this call always succeeds)
-		ret = append(ret, r) // add message receipt
+const (
+	// TODO: reduce these when expected gas costs are known
+	sumbitElectionPostGasLimit = 1e18
+	payBlockRewardGasLimit     = 1e1
+	cronTickGasLimit           = 1e18
+)
+
+// Applies all the message in a tipset, along with implicit block- and tipset-specific state
+// transitions.
+func (vmi *VMInterpreter_I) ApplyTipSetMessages(inTree st.StateTree, msgs TipSetMessages) (outTree st.StateTree, receipts []msg.MessageReceipt) {
+	outTree = inTree
+	seenMsgs := make(map[ipld.CID]struct{}) // CIDs of messages already seen once.
+	var r msg.MessageReceipt
+	for _, blk := range msgs.Blocks() {
+		// Pay block reward.
+		reward := _makeBlockRewardMessage(outTree, blk.Miner())
+		outTree, r = vmi.ApplyMessage(outTree, reward, blk.Miner())
+		if r.ExitCode() != exitcode.OK() {
+			panic("block reward failed")
+		}
+
+		// Process block miner's Election PoSt.
+		epost := _makeElectionPoStMessage(outTree, blk.Miner(), msgs.Epoch(), blk.PoStProof())
+		outTree, r = vmi.ApplyMessage(outTree, epost, blk.Miner())
+		if r.ExitCode() != exitcode.OK() {
+			panic("election post failed")
+		}
+
+		// Process messages from the block.
+		for _, m := range blk.Messages() {
+			_, found := seenMsgs[_msgCID(m)]
+			if found {
+				continue
+			}
+			outTree, r = vmi.ApplyMessage(outTree, m, blk.Miner())
+			receipts = append(receipts, r)
+			seenMsgs[_msgCID(m)] = struct{}{}
+		}
 	}
-	return compTree, ret
-}
-
-func _applyError(errCode exitcode.SystemErrorCode) msg.MessageReceipt {
-	// TODO: should this gasUsed value be zero?
-	// If nonzero, there is not guaranteed to be a nonzero gas balance from which to deduct it.
-	gasUsed := gascost.ApplyMessageFail
-	TODO()
-	return msg.MessageReceipt_MakeSystemError(errCode, gasUsed)
-}
 
-func _withTransferFundsAssert(tree st.StateTree, from addr.Address, to addr.Address, amount actor.TokenAmount) st.StateTree {
-	// TODO: assert amount nonnegative
-	retTree, err := tree.Impl().WithFundsTransfer(from, to, amount)
-	if err != nil {
-		panic("Interpreter error: insufficient funds (or transfer error) despite checks")
-	} else {
-		return retTree
+	// Invoke cron tick, attributing it to the miner of the first block.
```

Comment:
a bit late to the party, but just to note that applying `makeCronTickMessage` last means for any epoch-based timeouts, they occur at next epoch number. Ie say my sPost challenge comes in on epoch 400, I'll actually still be able to ePost at 400 and will only get the challenge on 401 if I don't post on 401 given the tick occurs last in the epoch.

It's functionally the same, just something to keep top of mind...